client/components/data/query-products-list/index.jsx

```diff
@@ -1,9 +1,16 @@
 import { useEffect } from 'react';
 import { useDispatch } from 'react-redux';
 import { requestProductsList } from 'calypso/state/products-list/actions';
-import { isProductsListFetching } from 'calypso/state/products-list/selectors';
+import { isProductsListFetching, getProductsList } from 'calypso/state/products-list/selectors';
 
 const request = ( props ) => ( dispatch, getState ) => {
+	if ( props.persist ) {
+		const productsList = getProductsList( getState() );
+		if ( productsList && Object.keys( productsList ).length > 0 ) {
```

Comment:
I am afraid that the `type` parameter will not allow us to use `Object.keys( productsList ).length > 0` since `type` might be used to fetch a specific segment of products which don't include marketplace (or other) products

https://github.com/Automattic/wp-calypso/blob/6a243bcf5a218680ac6480c919a078e9d2debb7a/client/my-sites/plans/jetpack-plans/selector.tsx#L215

this would result in not having the marketplace products in state and the query not running.

Steps to repro:
- Delete indexed db `productlist` key
- visit `/plans/[DOMAIN]` of a Jetpack site
- visit `/plugins/[DOMAIN]` of WPCOM site

<img width="1590" alt="SS 2022-01-21 at 14 39 49" src="https://user-images.githubusercontent.com/12430020/150528571-513d81ec-c57b-414e-8815-479b9aa70e0b.png">

Reply:
I've tried reproducing this with a Free plan as on the screenshot, but the jetpack request isn't fired for me on the `/plans/[DOMAIN]` page. Could you help me with which sites count as a Jetpack site?

Looking at the code where the query should be fired, this is a page where Jetpack plans are listed, right?
https://github.com/Automattic/wp-calypso/blob/980e64f34ca15ebf5afb2242655e0cd2615f8d89/client/my-sites/plans/jetpack-plans/selector.tsx#L215

Reply:
ahh, sorry. I mean self-hosted sites connected to WPCOM via jetpack.

Reply:
Thanks for the clarification @cpapazoglou! I was able to reproduce the issue this way. 👍 

I've added a new `type` field to the `productsList` state variable, and when the query is run, the type is being stored there. This way, when determining whether to fetch the query again, we can compare what has been previously saved to state.